coderd/workspaceapps/db.go

```diff
@@ -341,3 +363,192 @@ func (p *DBTokenProvider) authorizeRequest(ctx context.Context, roles *rbac.Subj
+
+		type additionalFields struct {
+			audit.AdditionalFields
+			SlugOrPort string `json:"slug_or_port,omitempty"`
+		}
+		appInfo := additionalFields{
+			AdditionalFields: audit.AdditionalFields{
+				WorkspaceOwner: aReq.dbReq.Workspace.OwnerUsername,
+				WorkspaceName:  aReq.dbReq.Workspace.Name,
+				WorkspaceID:    aReq.dbReq.Workspace.ID,
+			},
+		}
+		switch {
+		case aReq.dbReq.AccessMethod == AccessMethodTerminal:
+			appInfo.SlugOrPort = "terminal"
+		case aReq.dbReq.App.ID == uuid.Nil:
+			// If this isn't an app or a terminal, it's a port.
+			appInfo.SlugOrPort = aReq.dbReq.AppSlugOrPort
+		}
+
+		appInfoBytes, err := json.Marshal(appInfo)
+		if err != nil {
+			p.Logger.Error(ctx, "marshal additional fields failed", slog.Error(err))
+		}
+
+		userID := uuid.NullUUID{}
+		if aReq.apiKey != nil {
+			userID = uuid.NullUUID{Valid: true, UUID: aReq.apiKey.UserID}
+		}
+
+		var (
+			updatedIDs []uuid.UUID
+			sessionID  = uuid.Nil
+		)
+		err = p.Database.InTx(func(tx database.Store) error {
+			// nolint:gocritic // System context is needed to write audit sessions.
+			dangerousSystemCtx := dbauthz.AsSystemRestricted(ctx)
+
+			updatedIDs, err = tx.UpdateWorkspaceAppAuditSession(dangerousSystemCtx, database.UpdateWorkspaceAppAuditSessionParams{
+				AgentID:         aReq.dbReq.Agent.ID,
+				AppID:           uuid.NullUUID{Valid: aReq.dbReq.App.ID != uuid.Nil, UUID: aReq.dbReq.App.ID},
+				UserID:          userID,
+				Ip:              aReq.ip,
+				SlugOrPort:      appInfo.SlugOrPort,
+				UpdatedAt:       aReq.time,
+				StaleIntervalMS: p.WorkspaceAppAuditSessionTimeout.Milliseconds(),
+			})
+			if err != nil {
+				return xerrors.Errorf("update workspace app audit session: %w", err)
+			}
+			if len(updatedIDs) > 0 {
+				// Session is valid and got updated, no need to create a new audit log.
+				return nil
+			}
+
+			sessionID, err = tx.InsertWorkspaceAppAuditSession(dangerousSystemCtx, database.InsertWorkspaceAppAuditSessionParams{
+				AgentID:    aReq.dbReq.Agent.ID,
+				AppID:      uuid.NullUUID{Valid: aReq.dbReq.App.ID != uuid.Nil, UUID: aReq.dbReq.App.ID},
+				UserID:     userID,
+				Ip:         aReq.ip,
+				SlugOrPort: appInfo.SlugOrPort,
+				StartedAt:  aReq.time,
+				UpdatedAt:  aReq.time,
+			})
+			if err != nil {
+				return xerrors.Errorf("insert workspace app audit session: %w", err)
+			}
+
+			return nil
+		}, nil)
+		if err != nil {
+			p.Logger.Error(ctx, "update workspace app audit session failed", slog.Error(err))
+		}
+
+		if sessionID == uuid.Nil {
+			if sw.Status < 400 {
+				// Session was updated and no error occurred, no need to
+				// create a new audit log.
+				return
+			}
+			if len(updatedIDs) > 0 {
+				// Session was updated but an error occurred, we need to
+				// create a new audit log.
+				sessionID = updatedIDs[0]
+			} else {
+				// This shouldn't happen, but fall-back to request so it
+				// can be correlated to _something_.
+				sessionID = httpmw.RequestID(r)
+			}
+		}
```

Comment:
This control flow is a bit hard to follow. I think I disagree on how to handle an error. I think we should log and drop db errors.

If `err != nil`, this is either a db error (like a connection issue?) or an unique index violation? In either case, if we insert an audit log afterwards, it either will not succeed, or it will spam the audit log with entries. As every entry will have the same error.


I think `err != nil` should just log and `return` with 0 audit log entries. That also simplifies this control flow which is a bit hard to follow atm.

Reply:
I see what you mean. If err != nil it means there was an error communicating with the DB. My assumption was we'd still want to try to audit log in this case, but I can see how returning would be beneficial for clarity.

Returning at err != nil won't change the latter control flow, though. But we can change the behavior if deemed necessary. What going on there is:
- No active session -> audit log
- Active session updated, http status < 400 (success) -> skip
- Active session updated, http status >= 400 (error) -> audit log

Would you like to see some specific change in these scenarios?

Reply:
It's not immediately clear what the impact of these changes will be without some testing. Maybe we dogfood this and monitor the impact on audit logs, adjusting as necessary?

Reply:
> My fear is if the error is persistant (which I assume it would be for the same input data), then loading an app with ~50 asset files could trigger 50 audit logs.

That's a good point, so let's say you used to have access to an app, then reload it and it's a cached web-app that tries to load its resources -- you'd do a bunch of outbound requests and trigger the unauthorized log that many times. Yup, that checks out and probably not what we want.

This would adds complexity, but perhaps unauthorized logs shouldn't have the same 1h timeout as regular requests. I'm not sure how adamant we should be about catching abuse here.

Reply:
> It's not immediately clear what the impact of these changes will be without some testing. Maybe we dogfood this and monitor the impact on audit logs, adjusting as necessary?

Also a good suggestion. However I do think that unless we actively try to break it, we won't see much in dogfood.

Reply:
> However I do think that unless we actively try to break it, we won't see much in dogfood.

If we were able to run Coder as a `coder_app` it would definitely stress-test this 😄